### PR TITLE
Weight Initialization and Shared Layers

### PR DESCRIPTION
**Pull request by lilleswing**
Graph construction is not thread safe.  That is only one thread should be building the graph at a time.  In exchange for this now layers have much pretties names, "<__class__.name>_<number_initialized>" by default.

Create new function "shared" on Layer class as discussed https://github.com/deepchem/deepchem/issues/498.

Currently the only Layer to fill it is Dense.

### CODE REVIEW

File: deepchem/models/tensorgraph/layers.py [hunk 1]
```diff
@@ -77,30 +97,64 @@ def _create_tensor(self):
...
+                                                           num_outputs=self.out_channels,
+                                                           activation_fn=self.activation_fn,
+                                                           biases_initializer=self.biases_initializer(),
+                                                           weights_initializer=self.weights_initializer(),
+                                                           scope=self.scope_name,
+                                                           reuse=self.reuse,
+                                                           trainable=True)
+    self.out_tensor = tf.map_fn(dense_fn, parent.out_tensor)
+
+  def shared(self, in_layers):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: I think I get how this works (`tf.contrib.layers.fully_connected` looks up the variables in the provided `name_scope` and reuses them). That said, spooky action at a distance... I've never been a fan of TF's reliance on `namescopes`. Seems like a bad replacement for standard lexical scoping.

This isn't really a review comment. More me griping about Tensorflow :-)

File: deepchem/models/tensorgraph/layers.py [hunk 2]
```diff
@@ -609,3 +677,262 @@ def _create_tensor(self):
...
+
+    # Pack l (B, N) tensors into one (l, B, N) tensor
+    # Transpose to (B, N, l) for conv layer stacking
+    # done inside conv_layer loops to reduce transpose ops
+    # Final layer should be shape (N, B, l) to pass into tf.map_fn
+    # TODO (LESWING) batch norm
+    self.out_tensor = tf.stack(sym)
+    return self.out_tensor
+
+  def radialSymmetryFunction(self, R, rc, rs, e):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Could you change this to `def radial_symmetry_function(...)`? We haven't used camel-case so far in DeepChem.

File: deepchem/models/tensorgraph/layers.py [hunk 3]
```diff
@@ -609,3 +677,262 @@ def _create_tensor(self):
...
+
+    with tf.name_scope(None, "NbrRadialSymmetryFunction", [rc, rs, e]):
+      rc = tf.Variable(rc)
+      rs = tf.Variable(rs)
+      e = tf.Variable(e)
+      K = self.gaussianDistanceMatrix(R, rs, e)
+      FC = self.radialCutoff(R, rc)
+    return tf.multiply(K, FC)
+
+  def radialCutoff(self, R, rc):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same comment about names.

File: deepchem/models/tensorgraph/layers.py [hunk 4]
```diff
@@ -609,3 +677,262 @@ def _create_tensor(self):
...
+
+    """
+
+    T = 0.5 * (tf.cos(np.pi * R / (rc)) + 1)
+    E = tf.zeros_like(T)
+    cond = tf.less_equal(R, rc)
+    FC = tf.where(cond, T, E)
+    return FC
+
+  def gaussianDistanceMatrix(self, R, rs, e):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same comment about camel-case.

File: deepchem/models/tensorgraph/layers.py [hunk 5]
```diff
@@ -609,3 +677,262 @@ def _create_tensor(self):
...
+    Returns
+    -------
+      retval [B, N, M]: tf.Tensor
+        Gaussian distance matrix.
+
+    """
+
+    return tf.exp(-e * (R - rs)**2)
+
+  def distanceTensor(self, X, Nbrs, boxsize, B, N, M, d):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same comment here.

File: deepchem/models/tensorgraph/layers.py [hunk 6]
```diff
@@ -609,3 +677,262 @@ def _create_tensor(self):
...
+    example_nbrs = tf.unstack(nbr_indices, axis=0)
+    all_nbr_coords = []
+    for example, (example_tensor,
+                  example_nbr) in enumerate(zip(example_tensors, example_nbrs)):
+      nbr_coords = tf.gather(example_tensor, example_nbr)
+      all_nbr_coords.append(nbr_coords)
+    neighbors = tf.stack(all_nbr_coords)
+    return neighbors
+
+  def distanceMatrix(self, D):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: Same comment here.

File: deepchem/models/tensorgraph/layers.py [hunk 7]
```diff
@@ -77,30 +97,64 @@ def _create_tensor(self):
...
+      reuse=False,
+      **kwargs):
+    super(Dense, self).__init__(**kwargs)
     self.out_channels = out_channels
     self.out_tensor = None
     self.activation_fn = activation_fn
-    super(Dense, self).__init__(**kwargs)
+    self.biases_initializer = biases_initializer
+    self.weights_initializer = weights_initializer
+    self.time_series = time_series
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: You apply the dense network one at a time to each element of a "Batch".  This is generally done for time series data like videos.

File: deepchem/models/tensorgraph/layers.py [hunk 8]
```diff
@@ -49,6 +55,20 @@ def __eq__(x, y):
...
+    Parameters
+    ----------
+    in_layers: list tensor
+    List in tensors for the shared layer
+
+    Returns
+    -------
+    Layer
+    """
+    raise ValueError("Each Layer must implement shared for itself")
```
Comment (peastman) [inline deepchem/models/tensorgraph/layers.py]: Should be `NotImplementedError` rather than `ValueError`.

**Comment by peastman** [comment]
Looks nice!

File: deepchem/models/tensorgraph/layers.py (hunk 8 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: Changed

File: deepchem/models/tensorgraph/layers.py (hunk 6 above)
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: Names Changed

**Comment by rbharath** [comment]
@lilleswing and I chatted offline. The `time_series` functionality may be moved down to a subclass in a future PR, but we're good to go for now on this one. Will merge in once tests pass.

**Comment by rbharath** [comment]
Travis is passing now. LGTM.

**Pull request merged**